### PR TITLE
Implement type annotations using mypy

### PR DESCRIPTION
**Pull request by vddesai1998**
Closes https://github.com/HTTP-APIs/hydrus/issues/60

### Checklist
- [x] My branch is up-to-date with upstream/develop branch.
- [x] Everything works and tested for Python 3.5.2 and above.
- [x] I have created/claimed the issue that this PR resolves.

### CODE REVIEW

File: hydrus/app.py [hunk 1]
```diff
@@ -1,411 +1,434 @@
...
+                    except Exception as e:
+                        status_code, message = e.get_HTTP()
+                        return set_response_headers(jsonify(message), status_code=status_code)
+
+            return set_response_headers(jsonify({400: "Data is not valid"}), status_code=400)
+
+        abort(405)
+
+    def put(self, id_, type_):
+        """Add new object_ optional <id_> parameter using HTTP PUT."""
```
Comment (chrizandr) [inline hydrus/app.py]: I think this was missed

**Comment by vddesai1998** [comment]
How can I resolve this conflict?

**Comment by vddesai1998** [comment]
Tests were passing locally but once I made my branch up to date with develop authorization tests are failing

File: examples/drones/doc_gen.py [hunk 2]
```diff
@@ -3,8 +3,8 @@
 from hydrus.hydraspec.doc_writer import HydraDoc, HydraClass, HydraClassProp, HydraClassOp
 import json
 
-
 def doc_gen(API, BASE_URL):
+    # type: (str, str) -> HydraDoc
```
Comment (xadahiya) [inline examples/drones/doc_gen.py]: The type notations look fine but currently, these are nothing but comments. You need to implement static type checking and add mypy to the testing environment. 
See https://medium.com/@ageitgey/learn-how-to-use-static-type-checking-in-python-3-6-in-10-minutes-12c86d72677b and http://mypy.readthedocs.io/en/stable/getting_started.html for more info.

File: hydrus/data/db_models.py [hunk 3]
```diff
@@ -3,11 +3,12 @@
 from sqlalchemy import create_engine, ForeignKey
 from sqlalchemy.ext.declarative import declarative_base
 from sqlalchemy import Column, Integer, String
+from typing import Any
```
Comment (xadahiya) [inline hydrus/data/db_models.py]: Add typing to requirements.txt

File: examples/drones/doc_gen.py (hunk 2 above)
Comment (xadahiya) [inline examples/drones/doc_gen.py]: The type notations look fine but currently, these are nothing but comments. You need to implement static type checking and add mypy to the testing environment. 
See https://medium.com/@ageitgey/learn-how-to-use-static-type-checking-in-python-3-6-in-10-minutes-12c86d72677b and http://mypy.readthedocs.io/en/stable/getting_started.html for more info.

File: hydrus/data/legacy/generator.py [hunk 4]
```diff
@@ -1,6 +1,10 @@
 """Generate random objects for SubSystem classes."""
 
 import random
+from pyannotate_runtime import collect_types
```
Comment (xadahiya) [inline hydrus/data/legacy/generator.py]: Add pyannotate to requirements.txt

File: hydrus/tests/__init__.py [hunk 5]
```diff
@@ -1 +0,0 @@
-
```
Comment (xadahiya) [inline hydrus/tests/__init__.py]: Write tests for Static Type Checking using mypy.

**Comment by vddesai1998** [comment]
Thanks for reviewing. I will start implementing changes soon...

**Comment by vddesai1998** [comment]
And these are not mere comments but way to annotate types in mypy which is compatible with both python 2 and python 3. And pyannotate is not needed as requirement in our codebase. I used that open source tool to generate type annotations for our codebase . From now on contributors will annotate types while implementing new code. 

I will replace  those comment based annotations with python3.5 annotations syntax and write tests.

**Comment by xadahiya** [comment]
1. `And these are not mere comments but way to annotate types in mypy which is compatible with both python 2 and python 3` : As @Mec-iS  already specified, `As a generic rule, everything should work and be tested for Python 3.5.2 and above`. You don't need to make the codebase compatible with both python2 and python3.
2. `pyannotate is not needed as requirement in our codebase` : Doesn't matter if we need it in our codebase or not, if you're importing the module and it's not installed it will show an error. You should remove the import line if it's not required any more.

**Comment by vddesai1998** [comment]
Yeah I missed that. I should remove that unused import. Again thanks for reviewing

**Comment by Mec-iS** [comment]
As specified by @xadahiya we want the types annotations to be in the code not in comments.

**Comment by vddesai1998** [comment]
@Mec-iS @xadahiya Can you review it ? I made suggested changes , removed unused imports and  used python3 annotation syntax.

File: examples/drones/doc_gen.py (hunk 2 above)
Comment (vddesai1998) [reply examples/drones/doc_gen.py]: How should I add mypy to testing environment? As of now, I am running mypy in  hydrus directory with command  ```mypy  .``` to check the type errors. 
Tried mypy-pytest tool but it was too slow.
I think the best way would be to run mypy in the root directory as I am doing before pushing any changes.

**Comment by Mec-iS** [comment]
Please confirm that you run the `mypy` command and what it returns

**Comment by vddesai1998** [comment]
It returns errors about missing stub files. I tried to solve this but stub files for few libraries are not included in typeshed.
But when I run ```mypy --ignore-missing-imports``` on individual files it does not shows any errors.
I also ran ```mypy  .  --ignore-missing-imports``` in hydrus directory.

File: hydrus/app.py [hunk 6]
```diff
@@ -1,443 +1,444 @@
...
-        if get_authentication():
-            if request.authorization is None:
-                return failed_authentication()
-            else:
-                try:
-                    auth = check_authorization(request, get_session())
-                    if auth is False:
-                        return failed_authentication()
-                except Exception as e:
-                    status_code, message = e.get_HTTP()
```
Comment (xadahiya) [inline hydrus/app.py]: sketch 43

**Comment by Mec-iS** [comment]
@xadahiya I think we can make a custom exception class with a proper method and use that instead of `Exception`. You should open a related issue about this.

@chrizandr please provide indication to solve the error

**Comment by vddesai1998** [comment]
I can make it work with python 3.5.2  and other python 3 versions by making few changes but we will still get those attribute errors.

**Comment by vddesai1998** [comment]
In hydrus/exception.py we have some custom exceptions defined. How about using them?

File: hydrus/app.py (hunk 6 above)
Comment (vddesai1998) [reply hydrus/app.py]: Adapted to work with all python 3 versions

**Comment by vddesai1998** [comment]
@xadahiya @Mec-iS  I have made required changes.
I think we can merge this PR now.

Meanwhile, I can open another PR to solve Exception handling problem this way.
** From this**
```
def get(self, id_, type_):
        """GET object with id = id_ from the database."""
        if get_authentication():
            if request.authorization is None:
                return failed_authentication()
            else:
                try:
                    auth = check_authorization(request, get_session())
                    if auth is False:
                        return failed_authentication()
                except Exception as e:
                    status_code, message = e.get_HTTP()
                    return set_response_headers(jsonify(message), status_code=status_code)

        class_type = get_doc().collections[type_]["collection"].class_.title

        if checkClassOp(class_type, "GET"):

            try:
                response = crud.get(id_, class_type, api_name=get_api_name(), session=get_session())
                return set_response_headers(jsonify(hydrafy(response)))

            except Exception as e:
                status_code, message = e.get_HTTP()
                return set_response_headers(jsonify(message), status_code=status_code)

        abort(405)
```

**To this**

```
    def get(self, id_: int, type_: str) -> Response:
        """GET object with id = id_ from the database."""
        if get_authentication():
            if request.authorization is None:
                return failed_authentication()
            else:
                try:
                    auth = check_authorization(request, get_session())
                    if auth is False:
                        return failed_authentication()
                except UserNotFound as e:
                    status_code, message = e.get_HTTP()
                    return set_response_headers(jsonify(message), status_code=status_code)

        class_type = get_doc().collections[type_]["collection"].class_.title

        if checkClassOp(class_type, "GET"):

            try:
                response = crud.get(id_, class_type, api_name=get_api_name(), session=get_session())
                return set_response_headers(jsonify(hydrafy(response)))

            except ClassNotFound as e:
                status_code, message = e.get_HTTP() # type: ignore
                return set_response_headers(jsonify(message), status_code=status_code)

            except InstanceNotFound as e:
                status_code, message = e.get_HTTP() # type: ignore
                return set_response_headers(jsonify(message), status_code=status_code)

        abort(405)
```

Yes, I have imported those custom Exception classes from hydrus/data/exceptions.py

File: hydrus/app.py (hunk 6 above)
Comment (vddesai1998) [reply hydrus/app.py]: This is an issue with Exception handling. Type annotations are fine. As you can see in the screenshot there are no type errors.
I will open one more PR to solve that exception handling issue.
Thanks for reviewing!

**Comment by vddesai1998** [comment]
mypy

**Pull request merged**